File: core-commons/src/main/java/net/blueshell/common/enums/Role.java
package net.blueshell.common.enums;

import lombok.Getter;

import java.util.*;
import java.util.stream.Collectors;

public enum Role {
    GUEST("GUEST"),
    COMPANY("COMPANY"),
    MEMBER("MEMBER", GUEST),
    VEGAN("VEGAN"),
    COMMITTEE("COMMITTEE", MEMBER),
    BOARD("BOARD", COMMITTEE),
    ADMIN("ADMIN", BOARD),
    ;

    @Getter
    private final String reprString;
    private final Role[] inheritedRoles;

    Role(String reprString, Role... inheritedRoles) {
        this.reprString = reprString;
        this.inheritedRoles = inheritedRoles;
    }

    public boolean matchesRole(Role role) {
        return role == this || Arrays.stream(inheritedRoles).anyMatch(r -> r.matchesRole(role));
    }

    /**
     * Depth- (or breadth idk) first search for all inherited roles of this Role.
     */
    public Set<Role> getAllInheritedRoles() {
        Set<Role> res = new HashSet<>();
        res.add(this);
        ArrayDeque<Role> unexplored = new ArrayDeque<>(List.of(inheritedRoles));
        while (!unexplored.isEmpty()) {
            Role currentRole = unexplored.remove();
            res.add(currentRole);
            unexplored.addAll(Arrays.stream(currentRole.inheritedRoles).filter(role -> !res.contains(role)).collect(Collectors.toList()));
        }
        return res;
    }

    public Collection<Object> getAuthorities() {
        return Collections.singleton(getAllInheritedRoles());
    }

    public Object getName() {
        return this.reprString;
    }
}

File: api/src/main/kotlin/net/blueshell/api/shared/enums/Role.kt
package net.blueshell.api.shared.enums

import io.swagger.v3.oas.annotations.media.Schema
import java.util.*

@Schema(enumAsRef = true)
enum class Role(val reprString: String, vararg inheritedRoles: Role) {
    ANONYMOUS("ANONYMOUS"),
    VEGAN("VEGAN"),
    GUEST("GUEST", ANONYMOUS),
    COMPANY("COMPANY"),
    MEMBER("MEMBER", GUEST),
    COMMITTEE("COMMITTEE", MEMBER),
    BOARD("BOARD", COMMITTEE),
    TREASURER("TREASURER", BOARD),
    ADMIN("ADMIN", TREASURER),
    SYSTEM("SYSTEM", ADMIN),
    ;

    private val inheritedRoles: Array<Role> = arrayOf(*inheritedRoles)

    fun matchesRole(role: Role): Boolean {
        return role == this || inheritedRoles.any { it.matchesRole(role) }
    }

    val allInheritedRoles: MutableSet<Role>
        /**
         * Search for all inherited roles of this Role.
         */
        get() {
            val res: MutableSet<Role> = HashSet<Role>()
            res.add(this)
            val unexplored = ArrayDeque<Role>(inheritedRoles.toList())
            while (!unexplored.isEmpty()) {
                val currentRole = unexplored.remove()
                res.add(currentRole)
                for (role in currentRole.inheritedRoles) {
                    if (!res.contains(role)) {
                        unexplored.add(role)
                    }
                }
            }
            return res
        }

    val authorities: MutableCollection<Any>
        get() = this.allInheritedRoles.toMutableSet()

    companion object {
        fun allThatInherit(base: Role): MutableSet<Role> {
            val res = EnumSet.noneOf(Role::class.java)
            for (r in entries) {
                if (r.matchesRole(base)) {
                    res.add(r)
                }
            }
            return res
        }
    }
}
